bfvmm/src/hve/arch/intel_x64/vmi/vmi_op.cpp

```diff
@@ -0,0 +1,185 @@
+//
+// Copyright (C) 2020 Assured Information Security, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <hve/arch/intel_x64/vcpu.h>
+#include <hve/arch/intel_x64/emulation/cpuid.h>
+#include <hve/arch/intel_x64/vmi/vmi_op.h>
+
+#include <bfhypercall.h>
+#include <bfjson.h>
+
+#include <algorithm>
+
+namespace boxy::intel_x64
+{
+
+bool vmi_op_vmcall_dispatcher (vcpu_t *vcpu);
+
+vmi_op_handler::vmi_op_handler(
+    gsl::not_null<vcpu *> vcpu
+) :
+    m_vcpu{vcpu}
+{
+
+    // TODO:
+    //
+    // For now, dom0 is always the target of the introspecting VM
+    // and domU is always the introspecting VM.
+    // This will change once the vmi policy_op is implemented.
+
+    if(vcpu->is_domU()) {
+        vcpu->add_vmcall_handler({&vmi_op_handler::dispatch, this});
+        vcpu->add_cpuid_emulator(
+            bf_vmi_cpuid_leaf, {&vmi_op_handler::cpuid_vmi, this});
+    }
+}
+
+// Note:
+//
+// LibVMI currently makes hypercalls without changing the native ABI, except for setting RAX to the
+// hypercall opcode. Therefore, unlike other handlers, we have to follow SystemV's AMD64 ABI for
+// the calling convention to handle them.
+// As a reminder, the order is: RDI, RSI, RDX, RCX, R8, R9.
+
+void
+vmi_op_handler::vmi_op__memmap_ept(vcpu *vcpu)
+{
+    uint64_t addr = vcpu->rdi();
+    uint64_t gpa2 = vcpu->rsi();
+
+    auto [gpa1, unused1] = vcpu->gva_to_gpa(addr);
+    auto &&guest_map = get_domain(vcpu->domid())->ept();
+
+    if(guest_map.is_2m(gpa1))
+    {
+        bfdebug_info(0, "vmi_op: guest_map is 2m");
```

Comment:
Please remove these types of debug statements from your PR, or change the priority. Having them set to 0 is an issue.

Reply:
I changed the debug priority to 10. Let me know if that's fine.